### DISCUSSION TITLE
High CPU utilization using VAAPI on AMD GPU

### DISCUSSION THREAD

**Discussion by seanmikhaels** [immich-app/immich#8062] [General]
Hi All 

Seeing very high CPU utilization with VAAPI enabled. I've setup HWaccell.yml, enabled VAAPI in the main docker compose file and exposed /dev/dri during the setup. 

In Immich, I am using all default settings except for concurrent transcodes set to 1, bit rate at 2600K, transcode format h.264. I have an AMD 240GE ( yes only 2 Cores ), however, to compare transcoding a single 4k stream using FFMPEG with similar settings shows 10-15% utilization. 

Container Log shows successful encoding 

<img width="1368" alt="Screenshot 2024-03-18 at 11 02 37 PM" src="https://github.com/immich-app/immich/assets/39391683/7ca26d9f-39bd-44f6-b65d-51553e04e93f">

Almost no point in having HW acceleration enabled with such high CPU.

<img width="1390" alt="Screenshot 2024-03-18 at 11 17 21 PM" src="https://github.com/immich-app/immich/assets/39391683/f9f40019-a202-44aa-8913-7964ebbd4886">

**Comment by mertalev** [comment]
Decoding and tone-mapping is still done on CPU, so this is normal. With only two cores, the main benefit right now is faster transcoding rather than low CPU usage.

**Reply by seanmikhaels** [reply]
Thanks for the reply,  i was thinking about this right after i posted and totally forgot about tone mapping. The 4K comparison test i mentioned earlier was using HW decode as well, so that makes sense. Is there a reason or limitation why both HW Decode + Encode are not used?

**Reply by mertalev** [reply]
Mainly because the hardware tone-mapping is a lot more complicated. But we recently added it for one of the backends, so it should be possible to port that over to the other backends as well.

**Reply by redgyuf** [reply]
What does it mean "one of the backend". Is there a guess when it may available?

Currently im facing the same "issue" as OP with a i5-11400, I have the cpu cores close to full load while transcoding and the iGPU meanwhile is barely used.

This make the CPU run much hotter and had to set CPU pinning for immich so it wont eat up all the resources.

**Reply by mertalev** [reply]
It means RKMPP has end-to-end acceleration, but the others don't. There's no ETA for when this will be added.

**Reply by akostadinov** [reply]
Still not for VAAPI it seems with 115.0 :(

**Reply by mertalev** [reply]
Not yet, unfortunately. I won't be able to work on this soon, so PRs are welcome.

**Reply by akostadinov** [reply]
btw maybe it makes more sense to work on vulkan hardware encode/decode acceleration

https://www.phoronix.com/news/FFmpeg-Adds-Vulkan-Video-Encode

Seems to be promising as older cards are often limited with the supported formats. And Vulkan can bring new life to them.

Although fallback to software rendering needs to be done when format is unsupported.

P.S. ideally it should be configurable which formats use which acceleration type (or none). Because there are few cards to support all formats. And having things "sometimes" fail is not a viable option.

**Reply by mertalev** [reply]
I actually did work on this, first by changing almost completely to Vulkan in #9402, then reducing the scope in #9452 to exclude CPU and RKMPP. But Jellyfin devs in [this](https://github.com/jellyfin/jellyfin-ffmpeg/issues/391) thread emphasized that Vulkan wasn't mature enough yet.

I'm still interested in Vulkan and the situation is generally better now than it was at the time. I plan on adding it as an option, but I'd like to see what the reception is before really relying on it.

**Reply by mertalev** [reply]
> P.S. ideally it should be configurable which formats use which acceleration type (or none). Because there are few cards to support all formats.

If you have two devices and one supports a codec the other doesn't, it generally means that device is newer and you should just use that. It's rare for one device's codec support to not be a superset of the other, and newer devices generally offer better quality, capabilities and performance.

**Reply by akostadinov** [reply]
> If you have two devices and one supports a codec..

it is not about having multiple devices. But I have an old thinkcenter. Built-in radeon seems to only support h264 encoding and decoding of maybe everything except vp1 and 10bit hevc.

For encoding that is not an issue, because there is a single configured format. But for decoding, there is no guarantee which format will come.

So my remark was about when vulkan is implemented, that when some decoding format is not supported, would be useful to fallback to software rendering. Or maybe use whateevr native API is for supported formats and vulkan/software for the rest.

And thank you for working on this! I'm impressed so far with immich. Importing a friends collection of over 200GB now on this tiny server. Maybe I hit a small bug but will report separately.

**Reply by mertalev** [reply]
Oh, well it already does have software transcoding as a fallback.
